krm-functions/lib/ipallocation/v1alpha1/ipallocation.go

```diff
@@ -0,0 +1,436 @@
+	GetSelectorLabels() map[string]string
+	// GetSpecLabels returns the labelsfrom the spec
+	// if an error occurs or the attribute is not present an empty map[string]string is returned
+	GetSpecLabels() map[string]string
+	// GetCreatePrefix returns the create prefix from the spec
+	// if an error occurs or the attribute is not present false is returned
+	GetCreatePrefix() bool
+	// GetAllocatedPrefix returns the prefix from the status
+	// if an error occurs or the attribute is not present an empty string is returned
+	GetAllocatedPrefix() string
+	// GetAllocatedGateway returns the gateway from the status
+	// if an error occurs or the attribute is not present an empty string is returned
+	GetAllocatedGateway() string
+	// SetPrefixKind sets the prefixKind in the spec
+	SetPrefixKind(ipamv1alpha1.PrefixKind) error
+	// SetNetworkInstanceName sets the name of the networkInstance in the spec
+	SetNetworkInstanceName(s string) error
+	// SetAddressFamily sets the address family in the spec
+	SetAddressFamily(iputil.AddressFamily) error
+	// SetPrefix sets the prefix in the spec
+	SetPrefix(string) error
+	// SetPrefixLength sets the prefix length in the spec
+	SetPrefixLength(uint8) error
+	// SetIndex sets the index in the spec
+	SetIndex(uint32) error
+	// SetSelectorLabels sets the selector matchLabels in the spec
+	SetSelectorLabels(map[string]string) error
+	// SetSpecLabels sets the labels in the spec
+	SetSpecLabels(map[string]string) error
+	// SetCreatePrefix sets the create prefix in the spec
+	SetCreatePrefix(bool) error
+	// SetAllocatedPrefix sets the allocated prefix in the status
+	SetAllocatedPrefix(string) error
+	// SetAllocatedGateway sets the allocated gateway in the status
+	SetAllocatedGateway(string) error
+	// DeleteAddressFamily deletes the address family from the spec
+	DeleteAddressFamily() error
+	// DeletePrefix deletes the prefix from the spec
+	DeletePrefix() error
+	// DeletePrefixLength deletes the prefix length from the spec
+	DeletePrefixLength() error
+	// DeleteIndex deletes the index from the spec
+	DeleteIndex() error
+	// DeleteSpecLabels deletes the labels from the spec
+	DeleteSpecLabels() error
+	// DeleteCreatePrefix deletes the create prefix from the spec
+	DeleteCreatePrefix() error
+	// DeleteAllocatedPrefix deletes the allocated prefix from the status
+	DeleteAllocatedPrefix() error
+	// DeleteAllocatedGateway deletes the allocated gateway from the status
+	DeleteAllocatedGateway() error
+}
+
+// NewFromKubeObject creates a new parser interface
+// It expects a *fn.KubeObject as input representing the serialized yaml file
+func NewFromKubeObject(o *fn.KubeObject) IPAllocation {
+	return &obj{
+		p: parser.NewFromKubeObject[*ipamv1alpha1.IPAllocation](o),
+	}
+}
+
+// NewFromYaml creates a new parser interface
+// It expects raw byte slice as input representing the serialized yaml file
+func NewFromYAML(b []byte) (IPAllocation, error) {
+	p, err := parser.NewFromYaml[*ipamv1alpha1.IPAllocation](b)
```

Comment:
The github.com/GoogleContainerTools/kpt-functions-sdk/go/fn SDK that we use parses all YAML files before calling our "main" function (actually called Run()). It wraps the parsed YAML nodes into a hierarchy of KubeObject and SubObject instances. In order to avoid duplicate parsing, parsing from a YAML file (byte slice) should always be replaced by converting from a KubeObject, e.g. in this case by using the NewFromKubeObject() method of this interface.